pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+ * the prolongator. Formula can be found in Eq. (3.8) in [1].
+ *
+ * Parameters:
+ * -----------
+ *      n_nodes : const int
+ *          Number of rows in A
+ *      Ap : const array<int>
+ *          Row pointer for matrix A
+ *      Aj : const array<int>
+ *          Column indices for matrix A
+ *      Ax : const array<float>
+ *          Data array for matrix A
+ *      Sp : const array<int>
+ *          Row pointer for SOC matrix, C
+ *      Sj : const array<int>
+ *          Column indices for SOC matrix, C
+ *      Sx : const array<float>
+ *          Data array for SOC matrix, C -- MUST HAVE VALUES OF A
+ *      splitting : const array<int>
+ *          Boolean array with 1 denoting C-points and 0 F-points
+ *      Pp : const array<int>
+ *          Row pointer for matrix P
+ *      Pj : array<int>
+ *          Column indices for matrix P
+ *      Px : array<float>
+ *          Data array for matrix P
+ *
+ * Returns:
+ * --------
+ * Nothing, Pj[] and Px[] modified in place.
+ *
+ * References:
+ * -----------
+ * [0] J. W. Ruge and K. Stüben, Algebraic multigrid (AMG), in : S. F.
+ *      McCormick, ed., Multigrid Methods, vol. 3 of Frontiers in Applied
+ *      Mathematics (SIAM, Philadelphia, 1987) 73–130.
+ *
+ * [1] "Distance-Two Interpolation for Parallel Algebraic Multigrid,"
+ *      H. De Sterck, R. Falgout, J. Nolting, U. M. Yang, (2008).
+ */
+template<class I, class T>
+void rs_standard_interpolation_pass2(const I n_nodes,
+                                     const I Ap[], const int Ap_size,
+                                     const I Aj[], const int Aj_size,
+                                     const T Ax[], const int Ax_size,
+                                     const I Sp[], const int Sp_size,
+                                     const I Sj[], const int Sj_size,
+                                     const T Sx[], const int Sx_size,
+                                     const I splitting[], const int splitting_size,
+                                     const I Pp[], const int Pp_size,
+                                           I Pj[], const int Pj_size,
+                                           T Px[], const int Px_size)
+{
+    for (I i = 0; i < n_nodes; i++) {
+        // If node i is a C-point, then set interpolation as injection
+        if(splitting[i] == C_NODE) {
+            Pj[Pp[i]] = i;
+            Px[Pp[i]] = 1;
+        } 
+        // Otherwise, use RS standard interpolation formula
+        else {
+
+            // Calculate denominator
+            T denominator = 0;
+
+            // Start by summing entire row of A
+            for (I mm = Ap[i]; mm < Ap[i+1]; mm++) {
+                denominator += Ax[mm];
+            }
+
+            // Then subtract off the strong connections so that you are left with 
+            // denominator = a_ii + sum_{m in weak connections} a_im
+            for (I mm = Sp[i]; mm < Sp[i+1]; mm++) {
+                if ( Sj[mm] != i ) {
```

Comment:
@bensworth I think this could be a bug.  I don't think that the strength of connection matrix holds entries of A.  In classical_strength of connection, we do things like `S = scale_rows_by_largest_entry(S)`, possible amalgamation. That is, S_ij \neq A_ij
Then, evolution, energy-based, distance (and others) will hold even more different information.  
The quick and dirty solution would be to create a Ctemp in classical/classical.py, to be used only by the interpolation routines, e.g.,
`Ctemp = C.copy()`
`Ctemp.data[:] = 1.0`
`Ctemp = Ctemp.multiply(A)`